spec/models/sample_spec.rb

```diff
@@ -167,4 +167,79 @@
       expect(sample.sample_metadata.supplier_name).to be_nil
     end
   end
+
+  context 'compound samples' do
+    let(:child_sample1) { create(:sample) }
+    let(:child_sample2) { create(:sample) }
+    let(:parent_sample) { create(:sample) }
+
+    # let variables are lazy loaded and we always want the relationships to exist
+    # even if we don't access the parent sample in the test.
+    before do
+      parent_sample.update(component_samples: [child_sample1, child_sample2])
+      child_sample1.reload
+      child_sample2.reload
+    end
+
+    it 'parent samples are able to query their component samples (children)' do
+      expect(parent_sample.component_samples).to match_array [child_sample1, child_sample2]
+    end
+
+    it 'child samples are able to query their compound samples (parents)' do
+      expect(child_sample1.compound_samples).to match_array [parent_sample]
+      expect(child_sample2.compound_samples).to match_array [parent_sample]
+    end
+
+    it 'removing a component sample removes both sides of the relationship' do
+      parent_sample.component_samples.delete(child_sample2)
+      parent_sample.save
+      child_sample2.reload
+
+      expect(parent_sample.component_samples).to match_array [child_sample1]
+      expect(child_sample1.compound_samples).to match_array [parent_sample]
+      expect(child_sample2.compound_samples).to be_empty
+    end
+
+    it 'removing a compound sample removes both sides of the relationship' do
+      child_sample1.compound_samples.delete(parent_sample)
+      parent_sample.reload
+
+      expect(parent_sample.component_samples).to match_array [child_sample2]
+      expect(child_sample1.compound_samples).to be_empty
+      expect(child_sample2.compound_samples).to match_array [parent_sample]
+    end
+
+    it 'component samples can belong to many compound samples' do
+      other_parent = create(:sample, component_samples: [child_sample1])
+      child_sample1.reload
+
+      expect(other_parent.component_samples).to match_array [child_sample1]
+      expect(child_sample1.compound_samples).to match_array [parent_sample, other_parent]
+      expect(child_sample2.compound_samples).to match_array [parent_sample]
+    end
+
+    it 'fails validation when a compound sample is assigned a compound sample' do
+      other_sample = create(:sample)
+      parent_sample.compound_samples = [other_sample]
+
+      expect(parent_sample.save).to be_falsey
+      expect(parent_sample.errors[:compound_samples]).to include('cannot exist when component samples also exist')
+      expect(parent_sample.errors[:component_samples]).to include('cannot exist when compound samples also exist')
+
+      expect(child_sample1.save).to be_falsey
+      expect(child_sample1.errors[:compound_samples]).to include('cannot themselves have further compound samples')
```

Comment:
I find these two lines a little bit strange. I would expect that adding an illegal compound sample to `parent_sample` would break the `parent_sample` record, but wouldn't affect the existing sample `child_sample1`.

Happy to be convinced though...?

Reply:
Yes I think perhaps there are too many validations on the Sample in the relationships.  I think (maybe) the only constraint needs to be that a sample doesn't have both component samples and compound samples.  It shouldn't need to validate two layers away from itself.  What do you think?

Reply:
I haven't thought about it hard, but yes my concern was about validating two layers away, that it would be confusing that you could break records other than the one you're touching. I know I've had a pain in Sequencescape before when some existing bad data meant I couldn't do a thing with a record that seemed unrelated to the bad data. I know it's not unrelated in this case, but I still think it might be overkill. I think checking a sample doesn't have both component and compound samples sounds like a good approach.